tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -0,0 +1,201 @@
+                            const char *data,
+                            size_t len) {
+  size_t ret;
+  _httpsResponse_t *_httpsResponse = (_httpsResponse_t *)(parser->data);
+  /* nondet_bool is required to get coverage. */
+  _httpsResponse->foundHeaderField = nondet_bool();
+  _httpsResponse->parserState = PARSER_STATE_BODY_COMPLETE;
+  return ret;
+}
+
+IotNetworkError_t IotNetworkInterfaceCreate( void * pConnectionInfo,
+               void * pCredentialInfo,
+               void * pConnection ) {
+    *(char **)pConnection = malloc(1); /* network connection is opaque */
+    IotNetworkError_t error;
+    return error;
+}
+
+size_t IotNetworkInterfaceSend( void * pConnection,
+           const uint8_t * pMessage,
+           size_t messageLength ) {
+    size_t size;
+    __CPROVER_assume(size <= messageLength);
+    return size;
+}
+
+IotNetworkError_t IotNetworkInterfaceClose( void * pConnection ) {
+    IotNetworkError_t error;
+    return error;
+}
+
+size_t IotNetworkInterfaceReceive( void * pConnection,
+        uint8_t * pBuffer,
+        size_t bytesRequested ) {
+    uint8_t byte;
+    __CPROVER_array_copy(pBuffer,&byte);
+
+    size_t size;
+    __CPROVER_assume(size <= bytesRequested);
+    return size;
+}
+
+IotNetworkError_t IotNetworkInterfaceCallback( void * pConnection,
+                 IotNetworkReceiveCallback_t
+                 receiveCallback,
+                 void * pContext ) {
+    IotNetworkError_t error;
+    return error;
+}
+
+IotNetworkError_t IotNetworkInterfaceDestroy( void * pConnection ) {
+  IotNetworkError_t error;
+  return error;
+}
+
+IotNetworkInterface_t IOTNI = {
+  .create = IotNetworkInterfaceCreate,
+  .close = IotNetworkInterfaceClose,
+  .send = IotNetworkInterfaceSend,
+  .receive = IotNetworkInterfaceReceive,
+  .setReceiveCallback = IotNetworkInterfaceCallback,
+  .destroy = IotNetworkInterfaceDestroy
+};
+
+/* Models the Network Interface. */
+IotNetworkInterface_t *newNetworkInterface() {
+  return &IOTNI;
+}
+
+/* Creates a Connection Info and assigns memory accordingly. */
+IotHttpsConnectionInfo_t * newConnectionInfo() {
+  IotHttpsConnectionInfo_t * pConnInfo = safeMalloc(sizeof(IotHttpsConnectionInfo_t));
+  if(pConnInfo) {
+    pConnInfo->pNetworkInterface = newNetworkInterface();
+    /* The +1 is required for coverage */
+    __CPROVER_assume(pConnInfo->addressLen >= 0 && pConnInfo->addressLen <= IOT_HTTPS_MAX_HOST_NAME_LENGTH + 1);
+    pConnInfo->pAddress = safeMalloc(pConnInfo->addressLen);
+    /* The +1 is required for coverage */
+    __CPROVER_assume(pConnInfo->alpnProtocolsLen >= 0 && pConnInfo->alpnProtocolsLen <= IOT_HTTPS_MAX_ALPN_PROTOCOLS_LENGTH + 1);
+    pConnInfo->pAlpnProtocols = safeMalloc(pConnInfo->alpnProtocolsLen);
+    pConnInfo->pCaCert = malloc(sizeof(uint32_t));
+    pConnInfo->pClientCert = malloc(sizeof(uint32_t));
+    pConnInfo->pPrivateKey = malloc(sizeof(uint32_t));
+    pConnInfo->userBuffer.pBuffer = safeMalloc(sizeof(_connHandle_t));
```

Comment:
We can't use "connectionUserBufferMinimumSize" here instead of _connHandle_t?

Reply:
If we assign buffer instead of exactly specifying structure, CBMC takes too much time. Hence, we modeled the buffer with _connHandle_t structure which we have defined before.

Reply:
Ah ok missed the _connHandle_t structure that you created at the start of this file, the name looked so close to _httpsConnection_t.